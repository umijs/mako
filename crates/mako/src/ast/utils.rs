use swc_core::common::{Mark, DUMMY_SP};
use swc_core::ecma::ast::{
    CallExpr, Callee, Expr, ExprOrSpread, Ident, IdentName, Import, Lit, MemberExpr, MemberProp,
    MetaPropExpr, MetaPropKind, Module, ModuleItem,
};

use crate::module::{ModuleAst, ModuleSystem};

pub fn is_remote_or_data(url: &str) -> bool {
    let lower_url = url.to_lowercase();
    // ref:
    // https://developer.mozilla.org/en-US/docs/Web/CSS/url
    // https://www.ietf.org/rfc/rfc3986
    lower_url.starts_with("http://")
        || lower_url.starts_with("https://")
        || lower_url.starts_with("data:")
        || lower_url.starts_with("//")
}

pub fn is_remote_or_data_or_hash(url: &str) -> bool {
    let lower_url = url.to_lowercase();
    is_remote_or_data(url)
        // css url() should not resolve hash only url
        // e.g. fill: url(#gradient)
        || lower_url.starts_with('#')
}

pub fn remove_first_tilde(url: String) -> String {
    if let Some(stripped) = url.strip_prefix('~') {
        // ~/ is the case when use ~ as alias or folder
        if url.starts_with("~/") {
            url
        } else {
            stripped.to_string()
        }
    } else {
        url
    }
}

pub fn is_esm(module: &Module) -> bool {
    module
        .body
        .iter()
        .any(|item| matches!(item, ModuleItem::ModuleDecl(_)))
}

pub fn is_dynamic_import(call_expr: &CallExpr) -> bool {
    matches!(&call_expr.callee, Callee::Import(Import { .. }))
}

pub fn is_commonjs_require(call_expr: &CallExpr, unresolved_mark: &Mark) -> bool {
    if let Some(ident) = get_call_expr_ident(call_expr) {
        is_ident_undefined(ident, "require", unresolved_mark)
        // TODO: remove this, it's special logic
        || is_ident(ident, "__mako_require__")
    } else {
        false
    }
}

pub fn get_call_expr_ident(call_expr: &CallExpr) -> Option<&Ident> {
    if let Callee::Expr(box Expr::Ident(ident)) = &call_expr.callee {
        Some(ident)
    } else {
        None
    }
}

pub fn is_ident_undefined(ident: &Ident, sym: &str, unresolved_mark: &Mark) -> bool {
    ident.sym == *sym && ident.ctxt.outer() == *unresolved_mark
}

pub fn get_first_str_arg(call_expr: &CallExpr) -> Option<String> {
    if let Some(arg) = call_expr.args.first() {
        if let box Expr::Lit(Lit::Str(str_)) = &arg.expr {
            return Some(str_.value.to_string());
        }
    }
    None
}

pub fn is_ident(ident: &Ident, sym: &str) -> bool {
    ident.sym == *sym
}

pub fn is_import_meta_url(expr: &Expr) -> bool {
    matches!(
        expr,
        Expr::Member(MemberExpr {
            obj:
                box Expr::MetaProp(MetaPropExpr {
                    kind: MetaPropKind::ImportMeta,
                    ..
                }),
            prop:
                MemberProp::Ident(IdentName {
                    sym,
                    ..
                }),
            ..
        }) if sym == "url"
    )
}

pub fn id(s: &str) -> Ident {
    Ident {
        ctxt: Default::default(),
        span: DUMMY_SP,
        sym: s.into(),
        optional: false,
    }
}
pub fn member_prop(s: &str) -> MemberProp {
    MemberProp::Ident(IdentName {
        span: DUMMY_SP,
        sym: s.into(),
    })
}

pub fn promise_all(promises: ExprOrSpread) -> Expr {
    member_call(
        Expr::Ident(id("Promise")),
        member_prop("all"),
        vec![promises],
    )
}

pub fn member_call(obj: Expr, member_prop: MemberProp, args: Vec<ExprOrSpread>) -> Expr {
    Expr::Call(CallExpr {
        span: DUMMY_SP,
        ctxt: Default::default(),
        callee: Callee::Expr(Box::new(Expr::Member(MemberExpr {
            span: DUMMY_SP,
            obj: Box::new(obj),
            prop: member_prop,
        }))),
        args,
        type_args: None,
    })
}

pub fn require_ensure(source: String) -> Expr {
    member_call(
        Expr::Ident(id("__mako_require__")),
        MemberProp::Ident(id("ensure").into()),
        vec![ExprOrSpread {
            spread: None,
            expr: Box::new(Expr::Lit(Lit::Str(source.into()))),
        }],
    )
}

pub fn get_module_system(ast: &ModuleAst) -> ModuleSystem {
    match ast {
        ModuleAst::Script(module) => {
            let is_esm = module
                .ast
                .body
                .iter()
                .any(|s| matches!(s, ModuleItem::ModuleDecl(_)));
            if is_esm {
                ModuleSystem::ESModule
            } else {
                ModuleSystem::CommonJS
            }
        }
        crate::module::ModuleAst::Css(_) => ModuleSystem::Custom,
        crate::module::ModuleAst::None => ModuleSystem::Custom,
    }
}
