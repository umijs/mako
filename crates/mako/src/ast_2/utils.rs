use mako_core::base64::engine::general_purpose;
use mako_core::base64::Engine;
use mako_core::swc_common::DUMMY_SP;
use mako_core::swc_ecma_ast::{
    CallExpr, Callee, Expr, ExprOrSpread, Ident, Import, Lit, MemberExpr, MemberProp, MetaPropExpr,
    MetaPropKind, Module, ModuleItem,
};
use swc_core::common::Mark;

pub fn base64_encode<T: AsRef<[u8]>>(raw: T) -> String {
    general_purpose::STANDARD.encode(raw)
}

pub fn is_remote_or_data(url: &str) -> bool {
    let lower_url = url.to_lowercase();
    // ref:
    // https://developer.mozilla.org/en-US/docs/Web/CSS/url
    // https://www.ietf.org/rfc/rfc3986
    lower_url.starts_with("http://")
        || lower_url.starts_with("https://")
        || lower_url.starts_with("data:")
        || lower_url.starts_with("//")
}

pub fn is_remote_or_data_or_hash(url: &str) -> bool {
    let lower_url = url.to_lowercase();
    is_remote_or_data(url)
        // css url() should not resolve hash only url
        // e.g. fill: url(#gradient)
        || lower_url.starts_with('#')
}

pub fn remove_first_tilde(url: String) -> String {
    if let Some(stripped) = url.strip_prefix('~') {
        // ~/ is the case when use ~ as alias or folder
        if url.starts_with("~/") {
            url
        } else {
            stripped.to_string()
        }
    } else {
        url
    }
}

pub fn is_esm(module: &Module) -> bool {
    module
        .body
        .iter()
        .any(|item| matches!(item, ModuleItem::ModuleDecl(_)))
}

pub fn is_dynamic_import(call_expr: &CallExpr) -> bool {
    matches!(&call_expr.callee, Callee::Import(Import { .. }))
}

pub fn is_commonjs_require(call_expr: &CallExpr, unresolved_mark: &Mark) -> bool {
    if let Some(ident) = get_call_expr_ident(call_expr) {
        is_ident_undefined(ident, "require", unresolved_mark)
        // TODO: remove this, it's special logic
        || is_ident(ident, "__mako_require__")
    } else {
        false
    }
}

pub fn get_call_expr_ident(call_expr: &CallExpr) -> Option<&Ident> {
    if let Callee::Expr(box Expr::Ident(ident)) = &call_expr.callee {
        Some(ident)
    } else {
        None
    }
}

pub fn is_ident_undefined(ident: &Ident, sym: &str, unresolved_mark: &Mark) -> bool {
    ident.sym == *sym && ident.span.ctxt.outer() == *unresolved_mark
}

pub fn get_first_str_arg(call_expr: &CallExpr) -> Option<String> {
    if let Some(arg) = call_expr.args.first() {
        if let box Expr::Lit(Lit::Str(str_)) = &arg.expr {
            return Some(str_.value.to_string());
        }
    }
    None
}

pub fn is_ident(ident: &Ident, sym: &str) -> bool {
    ident.sym == *sym
}

pub fn is_import_meta_url(expr: &Expr) -> bool {
    matches!(
        expr,
        Expr::Member(MemberExpr {
            obj:
                box Expr::MetaProp(MetaPropExpr {
                    kind: MetaPropKind::ImportMeta,
                    ..
                }),
            prop:
                MemberProp::Ident(Ident {
                    sym,
                    ..
                }),
            ..
        }) if sym == "url"
    )
}

pub fn id(s: &str) -> Ident {
    Ident {
        span: DUMMY_SP,
        sym: s.into(),
        optional: false,
    }
}
pub fn member_prop(s: &str) -> MemberProp {
    MemberProp::Ident(Ident {
        span: DUMMY_SP,
        sym: s.into(),
        optional: false,
    })
}

pub fn promise_all(promises: ExprOrSpread) -> Expr {
    member_call(
        Expr::Ident(id("Promise")),
        member_prop("all"),
        vec![promises],
    )
}

pub fn member_call(obj: Expr, member_prop: MemberProp, args: Vec<ExprOrSpread>) -> Expr {
    Expr::Call(CallExpr {
        span: DUMMY_SP,
        callee: Callee::Expr(Box::new(Expr::Member(MemberExpr {
            span: DUMMY_SP,
            obj: Box::new(obj),
            prop: member_prop,
        }))),
        args,
        type_args: None,
    })
}

pub fn require_ensure(source: String) -> Expr {
    member_call(
        Expr::Ident(id("__mako_require__")),
        MemberProp::Ident(id("ensure")),
        vec![ExprOrSpread {
            spread: None,
            expr: Box::new(Expr::Lit(Lit::Str(source.into()))),
        }],
    )
}
