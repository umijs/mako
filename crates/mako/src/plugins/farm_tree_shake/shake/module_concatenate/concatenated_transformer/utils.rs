use super::super::ConcatenateContext;
use crate::module::ModuleId;

pub fn describe_export_map(ccn_ctx: &ConcatenateContext) -> String {
    let map = ccn_ctx.modules_exports_map.get(&ModuleId::from("mut.js"));

    if let Some(export_map) = map {
        let mut keys = export_map.keys().collect::<Vec<&String>>();
        keys.sort();
        let mut describe = String::new();
        keys.into_iter().for_each(|key| {
            let (id, sub) = export_map.get(key).unwrap();

            if let Some(field) = sub {
                describe.push_str(&format!("{} => {}.{}\n", key, id.sym, field));
            } else {
                describe.push_str(&format!("{} => {}\n", key, id.sym));
            }
        });

        describe.trim().to_string()
    } else {
        "None".to_string()
    }
}
