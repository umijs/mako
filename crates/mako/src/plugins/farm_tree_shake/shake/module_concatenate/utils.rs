use std::cmp::max;
use std::path::Path;

use heck::ToSnakeCase;

use crate::module::ModuleId;

pub fn uniq_module_prefix(module_id: &ModuleId) -> String {
    let path = Path::new(&module_id.id);
    let len = path.components().count() as i32;
    let mut skip = max(len - 3, 0);
    let mut p = path.components();
    while skip > 0 {
        p.next();
        skip -= 1;
    }

    format!(
        "__$m_{}",
        p.as_path().to_string_lossy().to_string().to_snake_case()
    )
}

// for define ast: `export { orig as exported }`
#[macro_export]
macro_rules! export_as {
    ( $orig:expr => $exported:expr ) => {
        mako_core::swc_ecma_ast::ExportSpecifier::Named(
            mako_core::swc_ecma_ast::ExportNamedSpecifier {
                span: DUMMY_SP,
                orig: mako_core::swc_ecma_ast::ModuleExportName::Ident($orig),
                exported: Some(mako_core::swc_ecma_ast::ModuleExportName::Ident(
                    $exported.clone(),
                )),
                is_type_only: false,
            },
        )
    };
}

#[macro_export]
macro_rules! export {
    ( $orig:expr ) => {
        mako_core::swc_ecma_ast::ExportSpecifier::Named(
            mako_core::swc_ecma_ast::ExportNamedSpecifier {
                span: DUMMY_SP,
                orig: mako_core::swc_ecma_ast::ModuleExportName::Ident($orig),
                exported: None,
                is_type_only: false,
            },
        )
    };
}

pub fn uniq_module_default_export_name(module_id: &ModuleId) -> String {
    format!("{}_0", uniq_module_prefix(module_id))
}

pub fn uniq_module_namespace_name(module_id: &ModuleId) -> String {
    format!("{}_ns", uniq_module_prefix(module_id))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn long_files_name() {
        let module_id = ModuleId::from("/a/very/very/deep/deep/module.js");
        let name = uniq_module_prefix(&module_id);
        assert_eq!(name, "__$m_deep_deep_module_js");
    }

    #[test]
    fn long_files_name_with_query() {
        let module_id = ModuleId::from("/a/very/very/deep/deep/module.js?aQuery");
        let name = uniq_module_prefix(&module_id);
        assert_eq!(name, "__$m_deep_deep_module_js_a_query");
    }

    #[test]
    fn just_file_name() {
        let module_id = ModuleId::from("module.js");
        let name = uniq_module_prefix(&module_id);
        assert_eq!(name, "__$m_module_js");
    }

    #[test]
    fn short_file_name() {
        let module_id = ModuleId::from("/module.js");
        let name = uniq_module_prefix(&module_id);
        assert_eq!(name, "__$m_module_js");
    }

    #[test]
    fn short_file_name_with_query() {
        let module_id = ModuleId::from("/module.js?asmodules");
        let name = uniq_module_prefix(&module_id);
        assert_eq!(name, "__$m_module_js_asmodules");
    }
}
