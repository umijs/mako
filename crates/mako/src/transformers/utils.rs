use swc_common::DUMMY_SP;
use swc_ecma_ast::{
    ArrowExpr, BlockStmtOrExpr, CallExpr, Callee, Expr, ExprOrSpread, Ident, Lit, MemberExpr,
    MemberProp, Pat,
};

pub fn id(s: &str) -> Ident {
    Ident {
        span: DUMMY_SP,
        sym: s.into(),
        optional: false,
    }
}
pub fn member_prop(s: &str) -> MemberProp {
    MemberProp::Ident(Ident {
        span: DUMMY_SP,
        sym: s.into(),
        optional: false,
    })
}

pub fn promise_all(promises: ExprOrSpread) -> Expr {
    member_call(
        Expr::Ident(id("Promise")),
        member_prop("all"),
        vec![promises],
    )
}

pub fn promise_resolve() -> Expr {
    member_call(Expr::Ident(id("Promise")), member_prop("resolve"), vec![])
}

pub fn member_call(obj: Expr, member_prop: MemberProp, args: Vec<ExprOrSpread>) -> Expr {
    Expr::Call(CallExpr {
        span: DUMMY_SP,
        callee: Callee::Expr(Box::new(Expr::Member(MemberExpr {
            span: DUMMY_SP,
            obj: Box::new(obj),
            prop: member_prop,
        }))),
        args,
        type_args: None,
    })
}

pub fn call(obj: Expr, args: Vec<ExprOrSpread>) -> Expr {
    Expr::Call(CallExpr {
        span: DUMMY_SP,
        callee: Callee::Expr(Box::new(obj)),
        args,
        type_args: None,
    })
}

pub fn arrow_fn(args: Vec<Pat>, body: Expr) -> Expr {
    Expr::Arrow(ArrowExpr {
        span: DUMMY_SP,
        params: args,
        body: Box::new(BlockStmtOrExpr::Expr(Box::new(body))),
        is_async: false,
        is_generator: false,
        type_params: None,
        return_type: None,
    })
}

pub fn require_ensure(source: String) -> Expr {
    member_call(
        Expr::Ident(id("require")),
        MemberProp::Ident(id("ensure")),
        vec![ExprOrSpread {
            spread: None,
            expr: Box::new(Expr::Lit(Lit::Str(source.into()))),
        }],
    )
}
