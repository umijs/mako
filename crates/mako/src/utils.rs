pub(crate) mod id_helper;
pub mod logger;
#[cfg(feature = "profile")]
pub mod profile_gui;
#[cfg(test)]
pub(crate) mod test_helper;
pub mod thread_pool;
pub mod tokio_runtime;

use std::path::Path;

use anyhow::{anyhow, Result};
use base64::engine::general_purpose;
use base64::Engine;
use cached::proc_macro::cached;
use regex::Regex;

pub fn base64_encode<T: AsRef<[u8]>>(raw: T) -> String {
    general_purpose::STANDARD.encode(raw)
}

pub fn base64_decode(bytes: &[u8]) -> Vec<u8> {
    general_purpose::STANDARD.decode(bytes).unwrap()
}

pub fn url_safe_base64_encode<T: AsRef<[u8]>>(raw: T) -> String {
    general_purpose::URL_SAFE.encode(raw)
}

pub fn url_safe_base64_decode(bytes: &[u8]) -> Vec<u8> {
    general_purpose::URL_SAFE.decode(bytes).unwrap()
}

pub trait ParseRegex {
    fn parse_into_regex(&self) -> Result<Option<Regex>>;
}

impl ParseRegex for Option<String> {
    fn parse_into_regex(&self) -> Result<Option<Regex>> {
        self.as_ref().map_or(Ok(None), |v| {
            Regex::new(v)
                .map(Some)
                .map_err(|_| anyhow!("Config Error invalid regex: {}", v))
        })
    }
}

pub fn process_req_url(public_path: &str, req_url: &str) -> Result<String> {
    let public_path = format!("/{}/", public_path.trim_matches('/'));
    if req_url.starts_with(&public_path) {
        return Ok(req_url[public_path.len() - 1..].to_string());
    }
    Ok(req_url.to_string())
}

pub(crate) fn get_app_info(root: &Path) -> (Option<String>, Option<String>) {
    let pkg_json_path = root.join("package.json");

    if pkg_json_path.exists() {
        let pkg_json = std::fs::read_to_string(pkg_json_path).unwrap();
        let pkg_json: serde_json::Value = serde_json::from_str(&pkg_json).unwrap();

        (
            pkg_json
                .get("name")
                .map(|name| name.as_str().unwrap().to_string()),
            pkg_json
                .get("version")
                .map(|name| name.as_str().unwrap().to_string()),
        )
    } else {
        (None, None)
    }
}

#[cached(key = "String", convert = r#"{ re.to_string() }"#)]
pub fn create_cached_regex(re: &str) -> Regex {
    Regex::new(re).unwrap()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_process_req_url() {
        assert_eq!(
            process_req_url("/public/", "/public/index.html").unwrap(),
            "/index.html"
        );
        assert_eq!(
            process_req_url("/public/foo/", "/public/foo/index.html").unwrap(),
            "/index.html"
        );
        assert_eq!(process_req_url("/", "/index.html").unwrap(), "/index.html");
        assert_eq!(
            process_req_url("/#/", "/#/index.html").unwrap(),
            "/index.html"
        );
    }
}
