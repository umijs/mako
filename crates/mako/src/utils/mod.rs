pub mod logger;
#[cfg(feature = "profile")]
pub mod profile_gui;
#[cfg(test)]
pub(crate) mod test_helper;
pub(crate) mod thread_pool;
pub mod tokio_runtime;

use anyhow::{anyhow, Result};
use base64::engine::general_purpose;
use base64::Engine;
use regex::Regex;

pub fn base64_encode<T: AsRef<[u8]>>(raw: T) -> String {
    general_purpose::STANDARD.encode(raw)
}

pub fn base64_decode(bytes: &[u8]) -> Vec<u8> {
    general_purpose::STANDARD.decode(bytes).unwrap()
}

pub trait ParseRegex {
    fn parse_into_regex(&self) -> Result<Option<Regex>>;
}

impl ParseRegex for Option<String> {
    fn parse_into_regex(&self) -> Result<Option<Regex>> {
        self.as_ref().map_or(Ok(None), |v| {
            Regex::new(v)
                .map(Some)
                .map_err(|_| anyhow!("Config Error invalid regex: {}", v))
        })
    }
}

pub fn process_req_url(public_path: &str, req_url: &str) -> Result<String> {
    let public_path = format!("/{}/", public_path.trim_matches('/'));
    if req_url.starts_with(&public_path) {
        return Ok(req_url[public_path.len() - 1..].to_string());
    }
    Ok(req_url.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_process_req_url() {
        assert_eq!(
            process_req_url("/public/", "/public/index.html").unwrap(),
            "/index.html"
        );
        assert_eq!(
            process_req_url("/public/foo/", "/public/foo/index.html").unwrap(),
            "/index.html"
        );
        assert_eq!(process_req_url("/", "/index.html").unwrap(), "/index.html");
        assert_eq!(
            process_req_url("/#/", "/#/index.html").unwrap(),
            "/index.html"
        );
    }
}
