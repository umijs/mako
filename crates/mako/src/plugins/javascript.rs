use std::sync::Arc;

use mako_core::anyhow::Result;
use mako_core::swc_atoms::js_word;
use mako_core::swc_common::collections::AHashSet;
use mako_core::swc_common::sync::Lrc;
use mako_core::swc_common::{self, Mark, GLOBALS};
use mako_core::swc_ecma_ast::{
    CallExpr, Callee, Expr, Id, Ident, Import, Lit, MemberExpr, MemberProp, MetaPropExpr,
    MetaPropKind, Module, ModuleDecl, NewExpr, Str,
};
use mako_core::swc_ecma_utils::collect_decls;
use mako_core::swc_ecma_visit::{Visit, VisitWith};

use super::css::is_url_ignored;
use crate::ast::build_js_ast;
use crate::compiler::Context;
use crate::load::{read_content, Content};
use crate::module::{Dependency, ModuleAst, ResolveType};
use crate::plugin::{Plugin, PluginDepAnalyzeParam, PluginLoadParam, PluginParseParam};

pub struct JavaScriptPlugin {}

impl Plugin for JavaScriptPlugin {
    fn name(&self) -> &str {
        "javascript"
    }

    fn load(&self, param: &PluginLoadParam, context: &Arc<Context>) -> Result<Option<Content>> {
        if param
            .task
            .is_match(vec!["js", "jsx", "ts", "tsx", "cjs", "mjs"])
        {
            if param.task.is_entry && param.task.request.has_query("hmr") {
                let port = &context.config.hmr_port.to_string();
                let host = &context.config.hmr_host.to_string();
                let host = if host == "0.0.0.0" { "127.0.0.1" } else { host };
                let content = format!(
                    "module.exports = require(\"{}\");",
                    param.task.request.path.as_str()
                );
                let content = format!(
                    "{}\n{}\n",
                    include_str!("../runtime/runtime_hmr_entry.js"),
                    content,
                )
                .replace("__PORT__", port)
                .replace("__HOST__", host);
                return Ok(Some(Content::Js(content)));
            }

            let content = read_content(param.task.request.path.as_str())?;
            return Ok(Some(Content::Js(content)));
        }
        Ok(None)
    }

    fn parse(&self, param: &PluginParseParam, context: &Arc<Context>) -> Result<Option<ModuleAst>> {
        if let Content::Js(content) = param.content {
            let ast = build_js_ast(&param.task.request.path, content, context)?;
            return Ok(Some(ModuleAst::Script(ast)));
        }
        Ok(None)
    }

    fn analyze_deps(
        &self,
        param: &mut PluginDepAnalyzeParam,
        context: &Arc<Context>,
    ) -> Result<Option<Vec<Dependency>>> {
        if let ModuleAst::Script(script) = param.ast {
            let mut visitor = DepCollectVisitor::new(script.unresolved_mark);
            GLOBALS.set(&context.meta.script.globals, || {
                script.ast.visit_with(&mut visitor);
                Ok(Some(visitor.dependencies))
            })
        } else {
            Ok(None)
        }
    }
}

struct DepCollectVisitor {
    bindings: Lrc<AHashSet<Id>>,
    dependencies: Vec<Dependency>,
    order: usize,
    unresolved_mark: Mark,
}

impl DepCollectVisitor {
    fn new(unresolved_mark: Mark) -> Self {
        Self {
            bindings: Default::default(),
            dependencies: vec![],
            // start with 1
            // 0 for swc helpers
            order: 1,
            unresolved_mark,
        }
    }
    fn bind_dependency(
        &mut self,
        source: String,
        resolve_type: ResolveType,
        span: Option<swc_common::Span>,
    ) {
        self.dependencies.push(Dependency {
            source,
            resolve_as: None,
            order: self.order,
            resolve_type,
            span,
        });
        self.order += 1;
    }
}

impl Visit for DepCollectVisitor {
    fn visit_module(&mut self, module: &Module) {
        self.bindings = Lrc::new(collect_decls(module));
        module.visit_children_with(self);
    }
    fn visit_module_decl(&mut self, n: &ModuleDecl) {
        match n {
            ModuleDecl::Import(import) => {
                if import.type_only {
                    return;
                }
                let src = import.src.value.to_string();
                self.bind_dependency(src, ResolveType::Import, Some(import.src.span));
            }
            ModuleDecl::ExportNamed(export) => {
                if let Some(src) = &export.src {
                    self.bind_dependency(
                        src.value.to_string(),
                        ResolveType::ExportNamed,
                        Some(src.span),
                    );
                }
            }
            ModuleDecl::ExportAll(export) => {
                let src = export.src.value.to_string();
                self.bind_dependency(src, ResolveType::ExportAll, Some(export.src.span));
            }
            _ => {}
        }
        // export function xxx() {} 里可能包含 require 或 import()
        n.visit_children_with(self);
    }
    fn visit_call_expr(&mut self, expr: &CallExpr) {
        if is_commonjs_require(expr, &self.unresolved_mark) {
            if let Some(src) = get_first_arg_str(expr) {
                self.bind_dependency(src, ResolveType::Require, Some(expr.span));
                return;
            }
        } else if is_dynamic_import(expr) {
            if let Some(src) = get_first_arg_str(expr) {
                self.bind_dependency(src, ResolveType::DynamicImport, Some(expr.span));
                return;
            }
        }
        expr.visit_children_with(self);
    }

    // Web workers
    fn visit_new_expr(&mut self, new_expr: &NewExpr) {
        if let Some(str) = resolve_web_worker(new_expr, self.unresolved_mark) {
            self.bind_dependency(str.value.to_string(), ResolveType::Worker, None);
        }

        new_expr.visit_children_with(self);
    }
}

pub fn resolve_web_worker(new_expr: &NewExpr, unresolved_mark: Mark) -> Option<&Str> {
    if !new_expr.args.as_ref().is_some_and(|args| !args.is_empty()) || !new_expr.callee.is_ident() {
        return None;
    }

    if let box Expr::Ident(Ident { span, sym, .. }) = &new_expr.callee {
        // `Worker` must be unresolved
        if sym == "Worker" && (span.ctxt.outer() == unresolved_mark) {
            let args = new_expr.args.as_ref().unwrap();

            // new Worker(new URL(''), base);
            if let Expr::New(new_expr) = &*args[0].expr {
                if !new_expr.args.as_ref().is_some_and(|args| !args.is_empty())
                    || !new_expr.callee.is_ident()
                {
                    return None;
                }

                if let box Expr::Ident(Ident { span, sym, .. }) = &new_expr.callee {
                    if sym == "URL" && (span.ctxt.outer() == unresolved_mark) {
                        // new URL(''); 仅第一个参数为字符串字面量, 第二个参数为 import.meta.url 时添加依赖
                        let args = new_expr.args.as_ref().unwrap();

                        if args.get(1).is_none() || !is_import_meta_url(&args.get(1).unwrap().expr)
                        {
                            return None;
                        }

                        if let box Expr::Lit(Lit::Str(ref str)) = &args[0].expr {
                            if !is_url_ignored(&str.value) {
                                return Some(str);
                            }
                        }
                    }
                }
            }
        }
    }

    None
}

pub fn is_import_meta_url(expr: &Expr) -> bool {
    matches!(
        expr,
        Expr::Member(MemberExpr {
            obj:
                box Expr::MetaProp(MetaPropExpr {
                    kind: MetaPropKind::ImportMeta,
                    ..
                }),
            prop:
                MemberProp::Ident(Ident {
                    sym: js_word!("url"),
                    ..
                }),
            ..
        })
    )
}

pub fn is_dynamic_import(call_expr: &CallExpr) -> bool {
    matches!(&call_expr.callee, Callee::Import(Import { .. }))
}

pub fn is_commonjs_require(call_expr: &CallExpr, unresolved_mark: &Mark) -> bool {
    if let Callee::Expr(box Expr::Ident(ident)) = &call_expr.callee {
        ident.sym == *"require" && is_native_ident(ident, unresolved_mark)
            || ident.sym == *"__mako_require__"
    } else {
        false
    }
}

pub fn is_native_ident(ident: &Ident, unresolved_mark: &Mark) -> bool {
    let outer = ident.span.ctxt.outer();

    outer == *unresolved_mark
}

pub fn get_first_arg_str(call_expr: &CallExpr) -> Option<String> {
    if let Some(arg) = call_expr.args.first() {
        if let box Expr::Lit(Lit::Str(str_)) = &arg.expr {
            return Some(str_.value.to_string());
        }
    }
    None
}
