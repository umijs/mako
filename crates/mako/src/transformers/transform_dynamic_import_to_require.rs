use mako_core::swc_ecma_ast::{Expr, ExprOrSpread, Lit};
use mako_core::swc_ecma_visit::{VisitMut, VisitMutWith};

use super::utils::{arrow_fn, call, id, member_call, member_prop, promise_resolve};
use crate::plugins::javascript::is_dynamic_import;

pub struct DynamicImportToRequire {}

// import('xxx') -> Promise.resolve().then(() => require('xxx'))
impl VisitMut for DynamicImportToRequire {
    fn visit_mut_expr(&mut self, expr: &mut Expr) {
        if let Expr::Call(call_expr) = expr {
            if is_dynamic_import(call_expr) {
                if let ExprOrSpread {
                    expr: box Expr::Lit(Lit::Str(ref mut source)),
                    ..
                } = &mut call_expr.args[0]
                {
                    // Promise.resolve().then(() => require('xxx'))
                    *expr = member_call(
                        promise_resolve(),
                        member_prop("then"),
                        vec![ExprOrSpread {
                            spread: None,
                            expr: Box::new(arrow_fn(
                                vec![],
                                call(
                                    Expr::Ident(id("require")),
                                    vec![ExprOrSpread {
                                        spread: None,
                                        expr: Box::new(Expr::Lit(Lit::Str(
                                            source.value.clone().into(),
                                        ))),
                                    }],
                                ),
                            )),
                        }],
                    );
                }
            }
        }
        expr.visit_mut_children_with(self);
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_basic_import() {
        crate::assert_display_snapshot!(transform(
            r#"
const testModule = import('test-module');
            "#,
        ));
    }

    #[test]
    fn test_chained_import() {
        crate::assert_display_snapshot!(transform(
            r#"
import('test-module').then(() => (
import('test-module-2')
));

Promise.all([
import('test-1'),
import('test-2'),
import('test-3'),
]).then(() => {});
            "#,
        ));
    }

    // TODO: support this?
    #[test]
    fn test_dynamic_import() {
        crate::assert_display_snapshot!(transform(
            r#"
import(MODULE);

let i = 0;
import(i++);

import(fn());

async () => import(await "x");

function* f() { import(yield "x"); }
            "#,
        ));
    }

    #[test]
    fn test_import_with_comment() {
        crate::assert_display_snapshot!(transform(
            r#"
import(/* test comment */ 'my-module');
import('my-module' /* test comment */ );
            "#,
        ));
    }

    // TODO: support this?
    #[test]
    fn test_template_argument() {
        crate::assert_display_snapshot!(transform(
            r#"
import(`1`);
import(tag`2`);
import(`3-${MODULE}`);
            "#,
        ));
    }

    fn transform(code: &str) -> String {
        let context = std::sync::Arc::new(Default::default());
        let mut visitor = super::DynamicImportToRequire {};
        crate::transformers::test_helper::transform_js_code(code, &mut visitor, &context)
    }
}
