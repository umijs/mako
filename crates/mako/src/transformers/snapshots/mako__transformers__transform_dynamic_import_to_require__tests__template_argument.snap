---
source: crates/mako/src/transformers/transform_dynamic_import_to_require.rs
expression: "transform(r#\"\nimport(`1`);\nimport(tag`2`);\nimport(`3-${MODULE}`);\n            \"#)"
---
import(`1`);
import(tag`2`);
import(`3-${MODULE}`);
